src/lib/sql/sql_translator.cpp

```diff
@@ -132,1176 +194,1031 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::translate_statement(const hsql::
-        stored_table_node->set_alias(alias);
-        return _translate_table_ref_alias(_validate_if_active(stored_table_node), table);
-      } else if (StorageManager::get().has_view(table.name)) {
-        node = StorageManager::get().get_view(table.name);
-        Assert(!_validate || node->subplan_is_validated(), "Trying to add non-validated view to validated query");
-      } else {
-        Fail(std::string("Did not find a table or view with name ") + table.name);
+  // Visitor that identifies aggregates and their arguments
+  const auto find_aggregates_and_arguments = [&](auto& sub_expression) {
+    if (sub_expression->type != ExpressionType::Aggregate) return true;
+
+    /**
+     * If the AggregateExpression has already been computed in a previous node (consider "x" in
+     * "SELECT x FROM (SELECT MIN(a) as x FROM t) AS y)", it doesn't count as a new Aggregate and is therefore not
+     * considered a "Aggregate" in the current SELECT list. Handling this as a special case seems hacky to me as well,
+     * but it's the best solution I can come up with right now.
+     */
+    if (_current_lqp->find_column_id(*sub_expression)) return false;
+
+    auto aggregate_expression = std::static_pointer_cast<AggregateExpression>(sub_expression);
+    if (aggregate_expression_set.emplace(aggregate_expression).second) {
+      aggregate_expressions.emplace_back(aggregate_expression);
+      for (const auto& argument : aggregate_expression->arguments) {
+        if (argument->requires_calculation()) {
+          if (pre_aggregate_expression_set.emplace(argument).second) {
+            pre_aggregate_expressions.emplace_back(argument);
+          }
+        }
       }
-      break;
-    case hsql::kTableSelect:
-      node = _translate_select(*table.select);
-      Assert(alias, "Every derived table must have its own alias");
-      node = _translate_table_ref_alias(node, table);
-      break;
-    case hsql::kTableJoin:
-      node = _translate_join(*table.join);
-      break;
-    case hsql::kTableCrossProduct:
-      node = _translate_cross_product(*table.list);
-      break;
-    default:
-      Fail("Unable to translate source table.");
-  }
+    }
 
-  node->set_alias(alias);
-  return node;
-}
+    return false;
+  };
 
-std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_where(const hsql::Expr& expr,
-                                                                 const std::shared_ptr<AbstractLQPNode>& input_node) {
-  DebugAssert(expr.isType(hsql::kExprOperator), "Filter expression clause has to be of type operator!");
+  // Identify all Aggregates and their arguments needed for SELECT and build the select_list_elements
+  // Each select_list_element is either an Expression or nullptr if the element is a Wildcard
+  std::vector<std::shared_ptr<AbstractExpression>> select_list_elements;
+  auto post_select_sql_identifier_context = std::make_shared<SQLIdentifierContext>(*_sql_identifier_context);
```

Comment:
Is the only reason we need to make a copy here to set the aliases in line 693? If so why can't we do it on the real context object?